core/arch/arm/plat-stm32mp1/plat_tzc400.c

```diff
@@ -33,91 +86,468 @@ static enum itr_return tzc_it_handler(struct itr_handler *handler __unused)
+					      reg_exclude->base + 1,
+					      reg->region.base,
+					      reg->region.top -
+					      reg->region.base + 1);
+		if (found)
+			break;
+	}
+
+	if (!found || !reg)
 		panic();
 
-	return region_cfg.base == base && region_cfg.top == (base + size - 1) &&
-	       region_cfg.sec_attr == TZC_REGION_S_RDWR &&
-	       region_cfg.ns_device_access == 0 &&
-	       region_cfg.filters == filters_mask;
+	if (reg_exclude->base == reg->region.base &&
+	    reg_exclude->top == reg->region.top) {
+		/* Remove this entry */
+		SLIST_REMOVE(&nsec_region_list, reg, tzc_region_non_sec, link);
+	} else if (reg_exclude->base == reg->region.base) {
+		reg->region.base = reg_exclude->top + 1;
+	} else if (reg_exclude->top == reg->region.top) {
+		reg->region.top = reg_exclude->base - 1;
+	} else {
+		struct tzc_region_non_sec *new_nsec =
+			calloc(1, sizeof(*new_nsec));
+
+		if (!new_nsec || !reg)
+			panic();
+
+		memcpy((void *)&new_nsec->region, (void *)&reg->region,
+		       sizeof(struct tzc_region_config));
+		reg->region.top = reg_exclude->base - 1;
+		new_nsec->region.base = reg_exclude->top + 1;
+		SLIST_INSERT_AFTER(reg, new_nsec, link);
+	}
+
+	return TEE_SUCCESS;
+}
+
+static void stm32mp_tzc_cfg_boot_region(struct tzc_device *tzc_dev)
+{
+	unsigned int idx = 0;
+	static struct tzc_region_config boot_region[] = {
+		{
+			.base = CFG_TZDRAM_START,
+			.top = CFG_TZDRAM_START + CFG_TZDRAM_SIZE - 1,
+			.sec_attr = TZC_REGION_S_RDWR,
+			.ns_device_access = 0,
+		},
+#ifdef CFG_CORE_RESERVED_SHM
+		{
+			.base = CFG_SHMEM_START,
+			.top = CFG_SHMEM_START + CFG_SHMEM_SIZE - 1,
+			.sec_attr = TZC_REGION_S_NONE,
+			.ns_device_access =
+				TZC_REGION_ACCESS_RDWR(STM32MP1_TZC_A7_ID),
+		}
+#endif
+	};
+
+	COMPILE_TIME_ASSERT(IS_PAGE_ALIGNED(CFG_TZDRAM_START));
+	COMPILE_TIME_ASSERT(IS_PAGE_ALIGNED(CFG_TZDRAM_SIZE));
+#ifdef CFG_CORE_RESERVED_SHM
+	COMPILE_TIME_ASSERT(IS_PAGE_ALIGNED(CFG_SHMEM_START));
+	COMPILE_TIME_ASSERT(IS_PAGE_ALIGNED(CFG_SHMEM_SIZE));
+#endif
+
+	stm32mp_tzc_region0(true);
+
+	stm32mp_tzc_reset_region(tzc_dev);
+
+	for (idx = 0; idx < ARRAY_SIZE(boot_region); idx++) {
+		TEE_Result res = TEE_ERROR_GENERIC;
+
+		boot_region[idx].filters =
+			filter_mask(tzc_dev->ddata->nb_filters);
+
+		res = stm32mp_tzc_region_append(tzc_dev, &boot_region[idx]);
+		if (res)
+			panic("Enable to configure core regions");
+
+		res = exclude_region_from_nsec(&boot_region[idx]);
+	}
+
+	/* Remove region0 access */
+	stm32mp_tzc_region0(false);
 }
 
-static TEE_Result init_stm32mp1_tzc(void)
+static TEE_Result add_node_memory_regions(struct tzc_device *tzc_dev,
+					  const void *fdt, int node)
+{
+	const fdt32_t *conf_list = NULL;
+	unsigned int nregions = 0;
+	unsigned int i = 0;
+	int len = 0;
+
+	conf_list = fdt_getprop(fdt, node, "memory-region", &len);
+	if (!conf_list)
+		return TEE_SUCCESS;
+
+	nregions = len / sizeof(uint32_t);
+	if (nregions > tzc_dev->ddata->nb_regions)
+		return TEE_ERROR_BAD_PARAMETERS;
```

Comment:
I suggest here a debug or error message: 
`EMSG("Too many regions defined in %s", fdt_get_name(fdt, node, NULL))`